src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+import Control.Lens hiding (levels)
+
+import qualified Data.Vector as V
+import qualified Data.Map.Strict as Map
+
+--------------------------
+-- Data Types
+--------------------------
+
+-- TODO: possibly allow more "atomic" expression types (besides variables) once
+--       we have some unboxed values
+
+-- TODO: switch back away from this
+newtype Arity = Arity Word32 deriving (Eq,Ord,Read,Show)
+
+data Anf
+  = AnfReturn !(V.Vector Variable) -- indices into the current env stack
+  | AnfLet !Arity -- TODO: switch back to !(V.Vector BinderInfo)
+           -- !(Maybe SourcePos)
+           !Rhs
+           !Anf
+  | AnfTailCall !App
+  deriving (Eq,Ord,Read,Show)
+
+data App
+  = AppFun !Variable
+           !(V.Vector Variable)
+  | AppPrim !PrimOpId !(V.Vector Variable)
+  | AppThunk !Variable
+  deriving (Eq,Ord,Read,Show)
+
+data Alloc
+  = AllocLit !Literal
+  | AllocLam !Arity -- TODO: switch back to !(V.Vector BinderInfo)
+             !Anf
+  | AllocThunk !Anf
+  deriving (Eq,Ord,Read,Show)
+
+data Rhs
+  = RhsAlloc !Alloc
+  | RhsApp !App
+  deriving (Eq,Ord,Read,Show)
+
+--------------------------
+-- Lowering to ANF
+--------------------------
+
+-- | The first slot in the most recent binder
+v0 :: Variable
+v0 = LocalVar $ LocalNamelessVar 0 $ BinderSlot 0
+
+-- | A temporary function to produce an 'Arity' for our ANF binding forms until
+-- we switch to using 'BinderInfo's in 'Anf' as well.
+arity :: V.Vector BinderInfo -> Arity
+arity binderInfos = Arity $ fromIntegral $ V.length binderInfos
+
+-- | A linear (single-use) reference to a(n existing/term or new/anf) binder. If
+-- two source variables refer to the same binder, they are tracked via separate
+-- 'AnfRef's.
+newtype AnfRef
+  = AnfRef { _refId :: Word64 }
+  deriving (Eq, Show, Ord)
+
+makeLenses ''AnfRef
+
+instance Enum AnfRef where
+  toEnum = AnfRef . toEnum
+  fromEnum = fromEnum . _refId
+
+-- | Bookkeeping structure that corresponds to a binder scope in the 'Term' AST.
+data BindingLevel
+  = BindingLevel { _levelRefs :: Map.Map AnfRef Variable
+                 -- ^ Linear references (to both existing/term and new/anf
+                 -- binders) to the 'Variable's they will be realized as when
+                 -- used. These variables are bumped as new 'AnfLet's are
+                 -- introduced.
+                 , _levelIntros :: Word32
+                 -- ^ Levels introduced since the last source binder. We keep
+                 -- this in addition to _levelRefs because we remove refs from
+                 -- the map once they've been used.
+                 , _levelIndirections :: Maybe (V.Vector Variable)
+                 -- ^ Set when the source binder(s) corresponding to this
+                 -- 'BindingLevel' point to an earlier variable. e.g. in
+                 -- translating 'Let's with 'V' or a 'Return' on the right-hand
+                 -- side, which will not have corresponding 'AnfLet's.
```

Comment:
This comment could be clarified: "number of binders introduced since the last source binder"